internal/cluster/cluster.go

```diff
@@ -668,6 +674,70 @@ func (m *Manager) UpdateLogsProgress(ctx context.Context, c *common.Cluster, pro
 	return err
 }
 
+func (m *Manager) UpdateInstallProgress(ctx context.Context, clusterID strfmt.UUID) error {
+	log := logutil.FromContext(ctx, m.log)
+
+	cluster, err := common.GetClusterFromDB(m.db, clusterID, common.UseEagerLoading)
+	if err != nil {
+		log.WithError(err).Error("Failed to get cluster from DB")
+		return err
+	}
+
+	var totalHostsDoneStages, totalHostsStages float64
+	for _, h := range cluster.Hosts {
+		stages := m.hostAPI.GetStagesByRole(h.Role, h.Bootstrap)
+		currentIndex := m.hostAPI.IndexOfStage(h.Progress.CurrentStage, stages)
+		totalHostsDoneStages += float64(currentIndex + 1)
+		totalHostsStages += float64(len(stages))
+	}
+
+	installingStagePercentage := int64((totalHostsDoneStages / totalHostsStages) * 100)
+	totalPercentage := int64(common.ProgressWeightPreparingForInstallationStage*float64(cluster.Progress.PreparingForInstallationStagePercentage) +
+		common.ProgressWeightInstallingStage*float64(installingStagePercentage))
+	updates := map[string]interface{}{
+		"progress_installing_stage_percentage": installingStagePercentage,
+		"progress_total_percentage":            totalPercentage,
+	}
+
+	return m.db.Model(&cluster).UpdateColumns(updates).Error
+}
+
+func (m *Manager) UpdateFinalizingProgress(ctx context.Context, db *gorm.DB, clusterID strfmt.UUID) error {
+	log := logutil.FromContext(ctx, m.log)
+
+	cluster, err := common.GetClusterFromDB(db, clusterID, common.UseEagerLoading)
+	if err != nil {
+		log.WithError(err).Error("Failed to get cluster from DB")
+		return err
+	}
+
+	var doneOperatorsCount float64
+	for _, o := range cluster.MonitoredOperators {
+
+		// built in operators must succeed
+		if o.OperatorType == models.OperatorTypeBuiltin && o.Status == models.OperatorStatusAvailable {
+			doneOperatorsCount++
+		}
+
+		// failing OLM operators will lead to dgraded cluster but are still considered as a progress
+		if o.OperatorType == models.OperatorTypeOlm && o.Status != models.OperatorStatusProgressing {
+			doneOperatorsCount++
+		}
+	}
+
+	finalizingStagePercentage := int64((doneOperatorsCount / float64(len(cluster.MonitoredOperators))) * 100)
+	totalPercentage := int64(common.ProgressWeightPreparingForInstallationStage*float64(cluster.Progress.PreparingForInstallationStagePercentage) +
+		common.ProgressWeightInstallingStage*float64(cluster.Progress.InstallingStagePercentage) +
+		common.ProgressWeightFinalizingStage*float64(finalizingStagePercentage))
```

Comment:
Currently - only finalizingStagePercentage should be less than 100

Reply:
Same response as https://github.com/openshift/assisted-service/pull/2208/files/9ab80aad5efd3881d8ec06c362c071a221118f90#r673715328.